File: fixtures/fake-scanner.ts
/**
 * A deterministic stand-in for `scanAndRedact`, shaped exactly like the
 * real result (`{ text, findings }`, `finding.action`), keyed on magic
 * substrings. Kept in sync by hand with `python/tests/fake_scanner.py`;
 * both implement the same four rules so the `*-cases.json` files in this
 * directory mean the same thing in either language.
 *
 * - text containing `BOOM` throws (a simulated core failure).
 * - text containing `BLOCK_ME` gets a `block` finding.
 * - text matching `SECRET_TOKEN_\d+` gets one `redact` finding over that
 *   span.
 * - text containing `WARN_ME` gets a `warn` finding (core leaves `warn`
 *   text untouched).
 * - anything else has no findings.
 */

import type { ScanResult, SecretAction, SecretFinding } from "@redact-secret/core";

function finding(action: SecretAction): SecretFinding {
  return {
    id: "finding-1",
    type: "generic_token",
    detector: "fake",
    confidence: action === "warn" ? "medium" : "high",
    obfuscation: "none",
    action,
    start: 0,
    end: 0,
  };
}

export function fakeScanAndRedact(text: string): ScanResult {
  if (text.includes("BOOM")) {
    throw new Error("simulated core failure - must never surface to a caller");
  }
  if (text.includes("BLOCK_ME")) {
    return { text: text.replace("BLOCK_ME", "<SECRET_1>"), findings: [finding("block")] };
  }
  const match = /SECRET_TOKEN_\d+/.exec(text);
  if (match) {
    const redacted = `${text.slice(0, match.index)}<SECRET_1>${text.slice(match.index + match[0].length)}`;
    return { text: redacted, findings: [finding("redact")] };
  }
  if (text.includes("WARN_ME")) {
    return { text, findings: [finding("warn")] };
  }
  return { text, findings: [] };
}

File: python/tests/fake_scanner.py
"""A deterministic stand-in for ``scan_and_redact``, shaped exactly like
the real ``ScanResult`` (``.text``, ``.findings``, each finding's
``.action``), keyed on magic substrings. Kept in sync by hand with
``fixtures/fake-scanner.ts``; both implement the same four rules so the
``*-cases.json`` files in the root ``fixtures/`` directory mean the same
thing in either language.

- text containing ``BOOM`` raises (a simulated core failure).
- text containing ``BLOCK_ME`` gets a ``block`` finding.
- text matching ``SECRET_TOKEN_\\d+`` gets one ``redact`` finding over
  that span.
- text containing ``WARN_ME`` gets a ``warn`` finding (core leaves
  ``warn`` text untouched).
- anything else has no findings.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field


@dataclass(frozen=True)
class FakeFinding:
    action: str
    id: str = "finding-1"
    type: str = "generic_token"
    detector: str = "fake"
    confidence: str = "high"
    obfuscation: str = "none"
    start: int = 0
    end: int = 0


@dataclass(frozen=True)
class FakeResult:
    text: str
    findings: list = field(default_factory=list)


_SECRET_TOKEN = re.compile(r"SECRET_TOKEN_\d+")


def fake_scan_and_redact(text: str, policy=None) -> FakeResult:
    if "BOOM" in text:
        raise RuntimeError("simulated core failure - must never surface to a caller")
    if "BLOCK_ME" in text:
        return FakeResult(text.replace("BLOCK_ME", "<SECRET_1>"), [FakeFinding("block")])
    match = _SECRET_TOKEN.search(text)
    if match:
        redacted = text[: match.start()] + "<SECRET_1>" + text[match.end() :]
        return FakeResult(redacted, [FakeFinding("redact")])
    if "WARN_ME" in text:
        return FakeResult(text, [FakeFinding("warn", confidence="medium")])
    return FakeResult(text, [])


class RecordingScanner:
    """``fake_scan_and_redact`` that also records each ``(text, policy)``
    call, so a test can assert what reached the core. Python-only; not part
    of the cross-language fixture contract."""

    def __init__(self) -> None:
        self.calls: list[tuple[str, object]] = []

    def __call__(self, text: str, policy=None) -> FakeResult:
        self.calls.append((text, policy))
        return fake_scan_and_redact(text, policy)
